xtext/org.icyphy.linguafranca/src/org/icyphy/generator/CppGenerator.xtend

```diff
@@ -671,43 +673,126 @@ class CppGenerator extends GeneratorBase {
     def templateLine(Reactor r) '''
         template<«FOR t: r.typeParms SEPARATOR ", "»class «t.toText»«ENDFOR»>
     '''
+    
+    /**
+     * Calculate the width of a multiport.
+     * FIXME: This currently
+     * throws an exception if the width depends on a parameter value.
+     * If the width depends on a parameter value, then this method
+     * will need to determine that parameter for each instance, not
+     * just class definition of the containing reactor.
+     */
+    def int calcPortWidth(VarRef port) {
+        val result = port.multiportWidth
+        if (result < 0) {
+            throw new Exception("Only multiport widths with literal integer values are supported for now.")
+        }
+        return result
+    }
 
-    def generate(Connection c) {
-        val leftContainer = c.leftPort.container
-        val rightContainer = c.rightPort.container
-        val leftPort = c.leftPort.variable as Port
-        val rightPort = c.rightPort.variable as Port
+    /**
+     * Calculate the width of a multiport.
+     * FIXME: This currently
+     * throws an exception if the width depends on a parameter value.
+     * If the width depends on a parameter value, then this method
+     * will need to determine that parameter for each instance, not
+     * just class definition of the containing reactor.
+     */
+    def int calcPortWidth(Port port) {
+        val result = port.widthSpec.width
+        if (result < 0) {
+            throw new Exception("Only multiport widths with literal integer values are supported for now.")
+        }
+        return result
+    }
 
-        if (leftContainer !== null && leftContainer.arraySpec !== null &&
-            rightContainer !== null && rightContainer.arraySpec !== null) {
-            return '''
-                for (unsigned i = 0; i < «leftContainer.name».size(); i++) {
-                  «leftContainer.name»[i].«leftPort.name».bind_to(&«rightContainer.name»[i].«rightPort.name»);
-                }
-            '''
-        } else if (leftContainer !== null && leftContainer.arraySpec !== null &&
-            rightPort.widthSpec !== null) {
-            return '''
-                for (unsigned i = 0; i < «leftContainer.name».size(); i++) {
-                  «leftContainer.name»[i].«leftPort.name».bind_to(&«c.rightPort.name»[i]);
+    def generate(Connection c) {
```

Comment:
It seems this function can easily be broken down into smaller parts. It's rather long now.